tools/clang/unittests/HLSL/OptimizerTest.cpp

```diff
@@ -259,3 +303,1152 @@ void OptimizerTest::OptimizerWhenSliceNThenOK(int optLevel, LPCSTR pText, LPCWST
+
+}
+
+TEST_F(OptimizerTest, OptimizerWhenPassedContainerPreservesViewId_GSDependent)
+{
+  ComparePSV0BeforeAndAfterOptimization(
+      R"(
+
+
+struct MyStructIn
+{
+//  float4 pos : SV_Position;
+  float4 a : AAA;
+  float2 b : BBB;
+  float4 c[3] : CCC;
+  float4 pos : SV_Position;
+};
+
+struct MyStructOut
+{
+  float4 pos : SV_Position;
+  float2 out_a : OUT_AAA;
+  uint d : SV_RenderTargetArrayIndex;
+};
+
+[maxvertexcount(18)]
+void main(triangleadj MyStructIn array[6], inout TriangleStream<MyStructOut> OutputStream0, uint viewid : SV_ViewID)
+{
+  float4 r = array[1].a + array[2].b.x + array[3].pos;
+  r += array[r.x].c[r.y].w + viewid;
+  MyStructOut output = (MyStructOut)0;
+  output.pos = array[r.x].a;
+  output.out_a = array[r.y].b;
+  output.d = array[r.x].a + 3;
+  OutputStream0.Append(output);
+  OutputStream0.RestartStrip();
+}
+
+
+)",
+      L"main", L"gs_6_5", true /*does use view id*/, 4);
+
+}
+
+using namespace llvm;
+using namespace hlsl;
+
+static std::unique_ptr<llvm::Module>
+    GetDxilModuleFromStatsBlobInContainer(LLVMContext & Context, IDxcBlob *pBlob) {
+  const char *pBlobContent =
+      reinterpret_cast<const char *>(pBlob->GetBufferPointer());
+  unsigned blobSize = pBlob->GetBufferSize();
+  const DxilContainerHeader *pContainerHeader =
+      IsDxilContainerLike(pBlobContent, blobSize);
+  const DxilPartHeader *pPartHeader =
+      GetDxilPartByType(pContainerHeader, DFCC_ShaderStatistics);
+  const DxilProgramHeader *pReflProgramHeader =
+      reinterpret_cast<const DxilProgramHeader *>(GetDxilPartData(pPartHeader));
+  (void)IsValidDxilProgramHeader(pReflProgramHeader, pPartHeader->PartSize);
+  const char *pReflBitcode;
+  uint32_t reflBitcodeLength;
+  GetDxilProgramBitcode((const DxilProgramHeader *)pReflProgramHeader,
+                        &pReflBitcode, &reflBitcodeLength);
+  std::string DiagStr;
+
+  return hlsl::dxilutil::LoadModuleFromBitcode(
+      llvm::StringRef(pReflBitcode, reflBitcodeLength), Context, DiagStr);
+}
+ 
+template<typename ResourceType>
+void CompareResources(const vector<unique_ptr<ResourceType>> &original,
+                      const vector<unique_ptr<ResourceType>> &optimized) {
+
+  VERIFY_ARE_EQUAL(original.size(), optimized.size());
+  for (size_t i = 0; i < original.size(); ++i) {
+    auto const &originalRes = original.at(i);
+    auto const &optimizedRes = optimized.at(i);
+    VERIFY_ARE_EQUAL(originalRes->GetClass(), optimizedRes->GetClass());
+    VERIFY_ARE_EQUAL(originalRes->GetGlobalName(), optimizedRes->GetGlobalName());
+    VERIFY_ARE_EQUAL(originalRes->GetGlobalSymbol()->getName(),
+                     optimizedRes->GetGlobalSymbol()->getName());
+    VERIFY_ARE_EQUAL(originalRes->GetHLSLType()->getTypeID(),
+                     optimizedRes->GetHLSLType()->getTypeID());
+    if (originalRes->GetHLSLType()->getTypeID() == Type::TypeID::StructTyID) {
+      VERIFY_ARE_EQUAL(originalRes->GetHLSLType()->getStructName(),
```

Comment:
`getStructName()` returns a llvm::StringRef.  This might be doing a pointer comparison on a `const char *` that this will implicitly cast to.  If execution ever reaches this code that is.

Reply:
I checked, and StringRef has an operator ==, so I think VERIFY_ARE_EQUAL is ok

Reply:
Ok, but what does it do if it's not equal?  I think the _STR variant is always better since it will write out the mismatching strings when it fails.

Reply:
Looks like VERIFY_ARE_EQUAL prints both values anyway, if they're different, and works with std::string (I tested that). Also there's no default conversion to char * for std::string or StringRef, so VERIFY_ARE_EQUAL_STR doesn't work without adding .c_str() anyway.

Example of comparison failure from  VERIFY_ARE_EQUAL operating on StringRef :

Error: Verify: AreEqual(originalPointedType->getStructName(), "Funny") [File: D:\DirectXShaderCompiler\tools\clang\unittests\HLSL\OptimizerTest.cpp, Function: CompareResources, Line: 1254]